### ISSUE TITLE
Unexpected, premature `File` drop in absence of `std::mem::forget`

### ISSUE THREAD

**Issue by rvolosatovs** [rust-lang/rust#104734]
I tried this code:

```rust
#[cfg(test)]
mod tests {
    use std::env::temp_dir;
    use std::fs::File;
    use std::io::{Read, Seek, Write};
    use std::os::unix::prelude::{AsRawFd, FromRawFd, OpenOptionsExt, RawFd};

    fn inner(fd: RawFd) {
        let mut file = unsafe { File::from_raw_fd(fd) };

        let mut s = String::new();
        file.read_to_string(&mut s).expect("failed to read");
        assert_eq!(s, "test");

        #[cfg(feature = "inner-forget")]
        std::mem::forget(file);
    }

    fn outer() {
        let mut file = File::options()
            .read(true)
            .write(true)
            .custom_flags(libc::O_TMPFILE)
            .open(temp_dir())
            .expect("failed to create temporary file");

        file.write(b"test").expect("failed to write to file");
        file.rewind().expect("failed to seek file");

        let fd = file.as_raw_fd();

        #[cfg(feature = "outer-forget")]
        std::mem::forget(file);

        inner(fd);
    }

    // NOTE: The failure is significantly easier to reproduce with three or more tests running.
    // It has been reproduced with two simultaneous tests.
    // It is not clear if the failure is reproducible with a single test case.

    #[test]
    fn one() {
        outer()
    }

    #[test]
    fn two() {
        outer()
    }

    #[test]
    fn three() {
        outer()
    }

    #[test]
    fn four() {
        outer()
    }
}
```

I expected to see this happen: success, even in absence of `std::mem::forget`

Instead, this happened: when `std::mem::forget` is not called in either `inner` or `outer` function, test sometimes fails on L28, before `AsRawFd::as_raw_fd` call. See https://github.com/rvolosatovs/fd-leak

Surprisingly, calling `std::mem::forget` from within `inner` also fixes the issue - I would not expect that to happen.

I would also expect `ManuallyDrop` to extend the lifetime of `file`:
```diff
diff --git a/src/lib.rs b/src/lib.rs
index fa94a93..809cdab 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -29,10 +29,14 @@ mod tests {
 
         let fd = file.as_raw_fd();
 
+        let file = std::mem::ManuallyDrop::new(file);
+
         #[cfg(feature = "outer-forget")]
         std::mem::forget(file);
 
         inner(fd);
+
+        std::mem::ManuallyDrop::into_inner(file);
     }
 
     // NOTE: The failure is significantly easier to reproduce with three or more tests running.
```
Which is not the case

### Meta

`rustc --version --verbose`:
```
rustc 1.65.0 (897e37553 2022-11-02)
binary: rustc
commit-hash: 897e37553bba8b42751c67658967889d11ecd120
commit-date: 2022-11-02
host: x86_64-unknown-linux-gnu
release: 1.65.0
LLVM version: 15.0.0
```

```
rustc 1.67.0-nightly (a28f3c88e 2022-11-20)
binary: rustc
commit-hash: a28f3c88e50a77bc2a91889241248c4543854e61
commit-date: 2022-11-20
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

<details><summary>Backtrace</summary>
<p>

```
RUST_BACKTRACE=full cargo test                  
    Finished test [unoptimized + debuginfo] target(s) in 0.02s
     Running unittests src/lib.rs (target/debug/deps/fd_leak-e25310d11ee23443)

running 4 tests
test tests::one ... ok
test tests::four ... ok
test tests::three ... ok
test tests::two ... FAILED

failures:

---- tests::two stdout ----
thread 'tests::two' panicked at 'failed to seek file: Os { code: 9, kind: Uncategorized, message: "Bad file descriptor" }', src/lib.rs:28:23
stack backtrace:
   0:     0x55eb7ea725a0 - std::backtrace_rs::backtrace::libunwind::trace::heb7a33f2fe285dd9
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55eb7ea725a0 - std::backtrace_rs::backtrace::trace_unsynchronized::hdd483bbded009146
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55eb7ea725a0 - std::sys_common::backtrace::_print_fmt::h66820ec1712881f4
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x55eb7ea725a0 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd2c196610029335c
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x55eb7ea95b4e - core::fmt::write::h2ca7d8a2085ff483
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/fmt/mod.rs:1208:17
   5:     0x55eb7ea6f4c5 - std::io::Write::write_fmt::hcad46088d384a7c3
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/io/mod.rs:1682:15
   6:     0x55eb7ea72365 - std::sys_common::backtrace::_print::h34ed3ab353b4c6af
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x55eb7ea72365 - std::sys_common::backtrace::print::h6c57f031c456c068
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x55eb7ea7413f - std::panicking::default_hook::{{closure}}::hec2c128303361806
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:267:22
   9:     0x55eb7ea73dfe - std::panicking::default_hook::hd13da34753ba960b
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:283:9
  10:     0x55eb7ea46707 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h5201491dfad1b856
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/alloc/src/boxed.rs:2032:9
  11:     0x55eb7ea46707 - test::test_main::{{closure}}::he18ce205e5b5b19a
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:135:21
  12:     0x55eb7ea7485d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h216311a9e3952096
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/alloc/src/boxed.rs:2032:9
  13:     0x55eb7ea7485d - std::panicking::rust_panic_with_hook::h73ea3535d824209e
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:692:13
  14:     0x55eb7ea745d7 - std::panicking::begin_panic_handler::{{closure}}::hd671ecd225a3dd76
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:579:13
  15:     0x55eb7ea72a4c - std::sys_common::backtrace::__rust_end_short_backtrace::hfb14ec03ea3b23c0
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:137:18
  16:     0x55eb7ea742f2 - rust_begin_unwind
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:575:5
  17:     0x55eb7ea149d3 - core::panicking::panic_fmt::h60d6645d301b37b5
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/panicking.rs:65:14
  18:     0x55eb7ea14c33 - core::result::unwrap_failed::ha4e2202a7759b47f
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/result.rs:1791:5
  19:     0x55eb7ea1526f - core::result::Result<T,E>::expect::hc79ac13a974c22a5
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/result.rs:1070:23
  20:     0x55eb7ea17c4d - fd_leak::tests::outer::h96195e8250fed54c
                               at /home/rvolosatovs/src/github.com/enarx/fd-leak/src/lib.rs:28:9
  21:     0x55eb7ea17cb6 - fd_leak::tests::two::hfc0baf4a4a6563e3
                               at /home/rvolosatovs/src/github.com/enarx/fd-leak/src/lib.rs:49:9
  22:     0x55eb7ea158e7 - fd_leak::tests::two::{{closure}}::h0f34485421b54f0b
                               at /home/rvolosatovs/src/github.com/enarx/fd-leak/src/lib.rs:48:14
  23:     0x55eb7ea17355 - core::ops::function::FnOnce::call_once::h9504b984ebcccf77
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/ops/function.rs:507:5
  24:     0x55eb7ea4b9cf - core::ops::function::FnOnce::call_once::hcdfabbc783fc005e
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/ops/function.rs:507:5
  25:     0x55eb7ea4b9cf - test::__rust_begin_short_backtrace::ha801a83b51fc7703
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:637:18
  26:     0x55eb7ea1de4c - test::run_test::{{closure}}::h6198e4affa99f9a6
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:628:30
  27:     0x55eb7ea1de4c - core::ops::function::FnOnce::call_once{{vtable.shim}}::h7f25ebb12fd78fc3
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/ops/function.rs:507:5
  28:     0x55eb7ea4a9b6 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hf5d0e8ff49b108f8
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/alloc/src/boxed.rs:2000:9
  29:     0x55eb7ea4a9b6 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::ha2b1c400f96a1a4b
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/panic/unwind_safe.rs:271:9
  30:     0x55eb7ea4a9b6 - std::panicking::try::do_call::hc71592039bdc38e9
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:483:40
  31:     0x55eb7ea4a9b6 - std::panicking::try::h08ecb4504cfedfc5
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:447:19
  32:     0x55eb7ea4a9b6 - std::panic::catch_unwind::hb97d2c3c5155970b
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panic.rs:137:14
  33:     0x55eb7ea4a9b6 - test::run_test_in_process::hca91b79341c56cd1
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:660:27
  34:     0x55eb7ea4a9b6 - test::run_test::run_test_inner::{{closure}}::h5ac461c4fd4af9c3
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:554:39
  35:     0x55eb7ea184b4 - test::run_test::run_test_inner::{{closure}}::h4349ca39f8d5f6f9
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/test/src/lib.rs:581:37
  36:     0x55eb7ea184b4 - std::sys_common::backtrace::__rust_begin_short_backtrace::h8aa33885e724cab6
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys_common/backtrace.rs:121:18
  37:     0x55eb7ea1ded5 - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::h4ca378ad3d438f78
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/thread/mod.rs:550:17
  38:     0x55eb7ea1ded5 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hcabdd1157f46f045
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/panic/unwind_safe.rs:271:9
  39:     0x55eb7ea1ded5 - std::panicking::try::do_call::h46f3095b7564eaa5
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:483:40
  40:     0x55eb7ea1ded5 - std::panicking::try::hee62ffac36e64677
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panicking.rs:447:19
  41:     0x55eb7ea1ded5 - std::panic::catch_unwind::h27ff2680cbabe863
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/panic.rs:137:14
  42:     0x55eb7ea1ded5 - std::thread::Builder::spawn_unchecked_::{{closure}}::h9c0ef1e4434dee3b
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/thread/mod.rs:549:30
  43:     0x55eb7ea1ded5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hc4cd48a559d98026
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/core/src/ops/function.rs:507:5
  44:     0x55eb7ea78bf3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha6912b1affeedc9f
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/alloc/src/boxed.rs:2000:9
  45:     0x55eb7ea78bf3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3ca4ddbfb732ff99
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/alloc/src/boxed.rs:2000:9
  46:     0x55eb7ea78bf3 - std::sys::unix::thread::Thread::new::thread_start::ha349c7e697b17ac9
                               at /rustc/a28f3c88e50a77bc2a91889241248c4543854e61/library/std/src/sys/unix/thread.rs:108:17
  47:     0x7f9d89fe7ff2 - start_thread
  48:     0x7f9d8a06abfc - clone3
  49:                0x0 - <unknown>


failures:
    tests::two

test result: FAILED. 3 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

</p>
</details>

**Comment by carbotaniuman** [comment]
The code as given is incorrect, and it makes sense as to why the non-determinism occurs, or why placing `mem::forget` fixes things.

To put it simply, `from_raw_fd` transfers ownership of the file descriptor, so what happens is that you create two `File`s that "own" the same file descriptor, and closes it twice. (This is equivalent to a double-free caused by misuse of `Box::into_raw` and `Box::from_raw`.

`cargo test` runs your tests in parallel, so what happens is something like this.

one(): opens new tempfile => fd 20
one(): close file in inner => close fd 20
two(): opens new tempfile => open fd 20
one(): close file in outer => close fd 20
two(): try to read from closed fd 20, fail

Basically, you should not use `from_raw_fd` with a FD from `as_fd`, only with one obtained from `into_raw_fd`, as that properly releases ownership. If you just want to borrow an FD without owning it/automatically closing it, consider https://doc.rust-lang.org/std/os/unix/io/struct.BorrowedFd.html.

**Comment by rvolosatovs** [comment]
> The code as given is incorrect, and it makes sense as to why the non-determinism occurs, or why placing `mem::forget` fixes things.
> 
> To put it simply, `from_raw_fd` transfers ownership of the file descriptor, so what happens is that you create two `File`s that "own" the same file descriptor, and closes it twice. (This is equivalent to a double-free caused by misuse of `Box::into_raw` and `Box::from_raw`.
> 
> `cargo test` runs your tests in parallel, so what happens is something like this.
> 
> one(): opens new tempfile => fd 20
> one(): close file in inner => close fd 20
> two(): opens new tempfile => open fd 20
> one(): close file in outer => close fd 20
> two(): try to read from closed fd 20, fail
> 
> Basically, you should not use `from_raw_fd` with a FD from `as_fd`, only with one obtained from `into_raw_fd`, as that properly releases ownership. If you just want to borrow an FD without owning it/automatically closing it, consider https://doc.rust-lang.org/std/os/unix/io/struct.BorrowedFd.html.

Makes sense, thanks.

I suspected this is what's going on from the beginning, but the inner `forget` fixing things threw me off.

`into_raw_fd` seems like the only way forward here, since the file descriptor has to be passed from one (Rust) binary to another in the "real" use case
